pkg/metrics/producers/scheduledcapacity/schedules.go

```diff
@@ -0,0 +1,500 @@
+			continue
+		}
+		if !(spec.matchesDay(iterBack)) {
+			iterBack.incrementDay(false)
+			continue
+		}
+		if !(spec.matchesHour(iterBack)) {
+			iterBack.incrementHour(false)
+			continue
+		}
+		if !(spec.matchesMinute(iterBack)) {
+			iterBack.incrementMinute(false)
+			continue
+		}
+		pastMatch = iterBack.convertToVolatileTime()
+	}
+	// only iterate for the next 5 years
+	for iterForward.isAfter(startTime) && iterForward.isBefore(endTime) {
+		if !(spec.matchesMonth(iterForward)) {
+			iterForward.incrementMonth(true)
+			continue
+		}
+		if !(spec.matchesDay(iterForward)) {
+			iterForward.incrementDay(true)
+			continue
+		}
+		if !(spec.matchesHour(iterForward)) {
+			iterForward.incrementHour(true)
+			continue
+		}
+		if !(spec.matchesMinute(iterForward)) {
+			iterForward.incrementMinute(true)
+			continue
+		}
+		futureMatch = iterForward.convertToVolatileTime()
+	}
+	if pastMatch.Inner.IsZero() {
+		return pastMatch, futureMatch, fmt.Errorf("unable to find a match in the past with the given ranges and crontab")
+	}
+	return pastMatch, futureMatch, nil
+}
+
+// A struct of fixed-length bitsets
+// The last value of the boolean is equivalent to field == *, meaning if the last index points to true, all values are valid
+type ScheduleSpec struct {
+	// 0-indexed
+	Minutes [61]bool
+	// 0-indexed
+	Hours [25]bool
+	// User input is 1-indexed
+	DaysOfMonth [32]bool
+	// User input is 1-indexed
+	Month [13]bool
+	// 0-indexed
+	DaysOfWeek [8]bool
+}
+
+func (s *ScheduleSpec) SetBitTrue(fieldType int, fieldIndex int) {
+	switch fieldType {
+	case 0:
+		s.Minutes[fieldIndex] = true
+	case 1:
+		s.Hours[fieldIndex] = true
+	case 2:
+		s.DaysOfMonth[fieldIndex] = true
+	case 3:
+		s.Month[fieldIndex] = true
+	case 4:
+		s.DaysOfWeek[fieldIndex] = true
+	}
+}
+
+// User input is 0-indexed
+func (s *ScheduleSpec) matchesMinute(iter *iterativeSchedule) bool {
+	return s.Minutes[iter.Minute] || s.Minutes[MINUTE_ASTERISK]
+}
+
+// User input is 0-indexed
+func (s *ScheduleSpec) matchesHour(iter *iterativeSchedule) bool {
+	return s.Hours[iter.Hour] || s.Hours[HOUR_ASTERISK]
+}
+
+// User input is 1-indexed
+func (s *ScheduleSpec) matchesMonth(iter *iterativeSchedule) bool {
+	return s.Month[iter.Month-1] || s.Month[MONTH_ASTERISK]
+}
+
+// User input of Day of Month is 1-indexed, Day of Week is 0-indexed
+func (s *ScheduleSpec) matchesDay(iter *iterativeSchedule) bool {
+
+	return (s.DaysOfWeek[iter.Dow] && s.DaysOfMonth[DOM_ASTERISK]) ||
+		(s.DaysOfMonth[iter.Dom-1] && s.DaysOfWeek[DOW_ASTERISK]) ||
+		(s.DaysOfWeek[iter.Dow] || s.DaysOfMonth[iter.Dom-1]) ||
+		(s.DaysOfWeek[DOW_ASTERISK] && s.DaysOfMonth[DOM_ASTERISK])
+}
+
+type iterativeSchedule struct {
+	Year   int
+	Minute int
+	Hour   int
+	Dom    int
+	// TODO Change to some struct that takes in string as well
+	Month int
+	// TODO Change to some struct that takes in string as well
+	Dow int
+}
+
+func (i *iterativeSchedule) convertToVolatileTime() *apis.VolatileTime {
```

Comment:
Consider renaming to VolatileTime(). It expresses the same thing. schedule.VolatileTime()